### ISSUE TITLE
[sdl2] Lightgun games don't work with the mouse

### ISSUE THREAD

**Issue by tmaul** [finalburnneo/FBNeo#744]
Currently the SDL2 port doesn't map the mouse correctly so lightgun games are "less playable". Can be fixed by manually editing the ini files but need to fix so it works by default or the Dinosaurs will win.

**Comment by dinkc64** [comment]
they must not win!!! :D

**Comment by grant2258** [comment]
what needs added to the config to fix ill start tracking the code if you like unless someone else is doing it.

**Comment by kwyxz** [comment]
As far as I can tell, this is still an issue in 2025. 

Config file clearly says mouseaxis 0 / 1 for gun movement : 

```
// FinalBurn Neo [DEBUG] v1.0.0.03 --- Config File for jpark (Jurassic Park (World, Rev A))

// --- Miscellaneous ----------------------------------------------------------

version 0x100003

analog  0x0100
cpu     0x0100



// --- Inputs -----------------------------------------------------------------

input  "P1 Coin"          switch 0x06
input  "P1 Start"         switch 0x02
input  "P1 Button 1"      switch 0x2C
input  "P1 Gun X"         mouseaxis 0
input  "P1 Gun Y"         mouseaxis 1
input  "P2 Coin"          switch 0x07
input  "P2 Start"         switch 0x03
input  "P2 Button 1"      switch 0x4080
input  "P2 Gun X"         joyaxis 0 0
input  "P2 Gun Y"         joyaxis 0 1
input  "Reset"            switch 0x3D
input  "Service Mode"     switch 0x3C
input  "Service 1"        switch 0x0A
input  "Service 2"        switch 0x0A
input  "Service 3"        switch 0x0A
input  "Tilt"             switch 0x14
input  "Dip A"            constant 0x0F
```

And when started, game shows mouse is present 

```
» ./fbneo -v jpark                                                                          kwyxz@ollie
FBNeo v1.0.0.03
Loading config from /home/kwyxz/.local/share/fbneo/config/fbneo.ini
*** Starting emulation of jpark - Jurassic Park (World, Rev A).
Cheat cpu-register INIT.
Graphics len: 800000, 1000000
PRG0: 00000, 0
Loading program (epr-16402a.ic8)... (OK)
PRG1: 100000, 1
Loading program (epr-16395.ic18)... (OK)
Loading program (epr-16394.ic9)... (OK)
Z80 : 00000, 3
Loading program (epr-16399.ic36)... (OK)
Z80 : 100000, 4
Loading program (mpr-16398.ic35)... (OK)
Z80 : 200000, 5
Loading program (mpr-16397.ic34)... (OK)
Z80 : 300000, 6
Loading program (mpr-16396.ic24)... (OK)
GFX0A: 00000, 7
Loading graphics (mpr-16404.ic14)... (OK)
Loading graphics (mpr-16403.ic5)... (OK)
GFX1A: 00000, 9
Loading graphics (mpr-16405.ic32)... (OK)
Loading graphics (mpr-16407.ic30)... (OK)
Loading graphics (mpr-16409.ic28)... (OK)
Loading graphics (mpr-16411.ic26)... (OK)
GFX1A: 800000, 13
Loading graphics (mpr-16406.ic31)... (OK)
Loading graphics (mpr-16408.ic29)... (OK)
Loading graphics (mpr-16410.ic27)... (OK)
Loading graphics (mpr-16412.ic25)... (OK)
BurnTimerInit: base index 0, #chips 2
SDLSoundInit (44100Hz) (6000FPS)
Game resolution: 320x224@60.000000
bbp: 32
setting logical size w: 320 h: 240
nVidImageWidth=320 nVidImageHeight=224 nVidImagePitch=1280
Malloc for video Ok 286720
p1 coin 5
p1 start 1
p1 fire 1 Z
mouse x-axis code 0x00
mouse y-axis code 0x100
p2 coin 6
p2 start 2
p2 fire 1 Joy 0 Button 0
p2 x-axis code 0x00
p2 y-axis code 0x100
reset F3
diag F2
service 9
service 9
service 9
tilt T
dip TAB
```

However mouse movement does not translate to gun movement in game at all.

**Comment by dinkc64** [comment]
I can't get the sdl2 version to compile here, using:

Windows, x64, gcc 15.2.0, commandline:
make sdl2 -j32 BUILD_X64_EXE=1

<img width="832" height="537" alt="Image" src="https://github.com/user-attachments/assets/eb80336f-3f65-446a-b33c-a48c000225b8" />

any ideas?  for fun, I removed all occurances of clear*opposite* from nes.cpp, and it didn't help.

**Comment by barbudreadmon** [comment]
Did you try removing "BUILD_X64_EXE=1" ? I don't think it's required (i don't use it on buildbot)

**Comment by dinkc64** [comment]
Yes, I originally compiled it without that and got the same issue.  
Hmm, maybe I could try compiling it without half the drivers and see how it goes.

**Comment by tmaul** [comment]
Hello! I just did a new SDL2 build on my Mac (the only machine I
have access to right now) and it built ok after deleting the obj directory.

Also, The coin-op Operation Wolf appears to work ok with mouse but lots of
other games fail. e.g. Operation Thunderbolt. I can have a try at debugging
but I hate MacOS and the dev tools are about as useful as doing a fart in a
hat.

**Comment by kwyxz** [comment]
> Also, The coin-op Operation Wolf appears to work ok with mouse but lots of
> other games fail. e.g. Operation Thunderbolt. 

That is a very good observation, I had not tried other games besides Jurassic Park, but I confirm that with my Linux SDL2 build the lightgun / mouse works fine with Operation Wolf indeed.

**Comment by dinkc64** [comment]
Oh this is good news - if some games work, buit some don't - that means it's probably a simple fix :)

Since I still can't compile sdl2 here, I'll try to fix the problem blindly

best regards,
- dink

**Comment by dinkc64** [comment]
Hi tmaul :)

**Comment by kwyxz** [comment]
> Since I still can't compile sdl2 here, I'll try to fix the problem blindly

Happy to help, if you need a tester.

**Comment by dinkc64** [comment]
@kwyxz sounds good :)

**Comment by dinkc64** [comment]
@tmaul can you help me try to compile this on windows with gcc?
make sdl2 :)

**Comment by dinkc64** [comment]
sorry guys, having a bad day here.

**Comment by barbudreadmon** [comment]
Hmmm, input-wise, the only difference i can see between opwolf and jpark/othunder is the number of players, so i'd expect there is some kind of conflict happening when mulltiple X/Y axis are listed ?
Let's ping @cjom about this ? I think he wrote most of the current mapping code in the sdl2 port.

**Comment by tmaul** [comment]
I’m going to try and get it building on my Windows PC later so I can have a proper look. Apple machines are not set up for development work at all, although it does build really very quickly compared to anything else I’ve ever compiled FBNeo on.

I seem to remember that last time I looked at this it was something to do with the input naming but I never got to the bottom of it.

**Comment by cjom** [comment]
If I recall correctly, sdl2 version only listens to game controllers events when mapping, maybe keyboard too. I'll try to look into that later today.

**Comment by dinkc64** [comment]
not exactly sure why it's needed, but linker flags "-Wl,--image-base -Wl,0x10000000" fixes the linking of executable on win*

https://github.com/finalburnneo/FBNeo/commit/c36790f900c175af67f1c6005cca31a1056209af

**Comment by tmaul** [comment]
On Mac It does complain that one of the code segments is too big for its
default linking so maybe it's something related to what you are seeing.
FBNeo has become too powerful

**Comment by dinkc64** [comment]
@cjom, any chance you could add a small UX: 
map home/end to go to the top & bottom of the list in the menus & input maps?
and escape to get out of the menu / go back to game regardless of menu depth

best regards,
- dink

**Comment by dinkc64** [comment]
@tmaul, too powerful, I like that :)

**Comment by cjom** [comment]
Just a quick look, but the function that sets controls is
https://github.com/finalburnneo/FBNeo/blob/c36790f900c175af67f1c6005cca31a1056209af/src/burner/sdl/input_sdl2.cpp#L134

declared in https://github.com/finalburnneo/FBNeo/blob/c36790f900c175af67f1c6005cca31a1056209af/src/burner/gamc.cpp#L10

in particular here:
https://github.com/finalburnneo/FBNeo/blob/c36790f900c175af67f1c6005cca31a1056209af/src/burner/gamc.cpp#L526

**Comment by tmaul** [comment]
I’ve not had much time to look at this but I think it’s quite likely that it’s only assigning the mouse to the last player as per what @barbudreadmon suggested. I’ll figure it out, just need more time.

**Comment by wn2000** [comment]
Fyi sdl2 only supports one mouse, despite the mouse event having a "which" field. Or multiple mouse devices control one global mouse cursor.

Might have something to do with this issue.
